src/vtpty/SshSession.cpp

```diff
@@ -0,0 +1,1168 @@
+            return;
+
+        if (!_p->sshClient.handshake())
+            return;
+
+        if (!_p->sshClient.authenticate(_username))
+        {
+            sshLog()("SSH agent based authentication failed. Trying again with password.");
+            if (!_p->sshClient.authenticate(_username, "winki"))
+                return;
+        }
+
+        if (!_p->sshClient.openChannel())
+            return;
+
+        if (SshClient::supportsAuthAgentForwarding())
+            _p->sshClient.requestAuthAgentForwarding(); // Continue on failure
+
+        if (!_p->sshClient.requestPty(_env.count("TERM") ? _env.at("TERM") : "",
+                                   _pageSize.columns.as<int>(),
+                                   _pageSize.lines.as<int>(),
+                                   _pixels.has_value() ? _pixels->width.as<int>() : 0,
+                                   _pixels.has_value() ? _pixels->height.as<int>() : 0))
+            return;
+
+        _p->sshClient.setEnvironment(_env);
+
+        if (!_p->sshClient.startShell())
+            return;
+
+        // _p->sshClient.setBlocking(false);
+    */
+}
+
+PtySlave& SshSession::slave() noexcept
+{
+    return *_ptySlave; // Can we find a way to avoid this?
+}
+
+void SshSession::close()
+{
+    setState(State::Closed);
+
+    if (_p->sshChannel)
+    {
+        libssh2_channel_send_eof(_p->sshChannel);
+        libssh2_channel_close(_p->sshChannel);
+        libssh2_channel_wait_closed(_p->sshChannel);
+    }
+
+    if (_p->sshSocket.is_open())
+        _p->sshSocket.close();
+}
+
+void SshSession::waitForClosed()
+{
+    // TODO: find better implementation (using polling)
+    while (!isClosed())
+        std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+}
+
+bool SshSession::isClosed() const noexcept
+{
+    return _p->sshSocket.is_closed() || _state == State::Closed || _state == State::Failure;
+}
+
+SshSession::ReadResult SshSession::read(crispy::buffer_object<char>& storage,
+                                        std::optional<std::chrono::milliseconds> timeout,
+                                        size_t size)
+{
+    auto injectLock = std::unique_lock { _injectMutex };
+
+    auto const isWaitSatisfied = [this]() {
+        return _state == State::Operational || !_injectedRead.empty();
+    };
+
+    while (!isWaitSatisfied())
+        _injectCV.wait(injectLock, isWaitSatisfied);
+
+    if (!_injectedRead.empty())
+    {
+        auto const nread = std::min(size, _injectedRead.size());
+        std::copy_n(_injectedRead.begin(), nread, storage.hotEnd());
+        _injectedRead.erase(0, nread);
+        return std::tuple { std::string_view { storage.hotEnd(), nread }, false };
+    }
+
+    if (_state == State::AuthenticatePasswordWaitForInput)
+    {
+        printf("xxxxxxx\n");
+        _p->wantsWaitForSocket = true;
+        waitForSocket(std::chrono::seconds(1));
+        std::this_thread::sleep_for(std::chrono::milliseconds(1'000));
```

Comment:
can we make this time a constant inside ssh session class?

Reply:
all timeouts and sleeps removed or replaced with condition variables.